### ISSUE TITLE
v4.16.2: `docker context list` gets "unexpected end of JSON input"

### ISSUE THREAD

**Issue by rfay** [docker/for-win#13180]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: EE17CD29-04CD-4EEB-A128-5EE34779FFD8/20230119214601

### Actual behavior

PS C:\Users\testbot> docker context list
unexpected end of JSON input

### Expected behavior

`docker context list` should give a list of docker contexts

This happens in PS and in WSL2 as well, same thing. I tested on two different Windows systems.
I tested on macOS and did not see this.

### Information

  - Windows Version: 11
  - Docker Desktop Version: 4.16.1
  - WSL2 or Hyper-V backend?  WSL2
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: No

### Output of `& "C:\Program Files\Docker\Docker\resources\com.docker.diagnose.exe" check`

Skipped since I haven't ever seen it work

### Steps to reproduce the behavior

On 4.16.1, `docker context list`

**Comment by rfay** [comment]
I should note that as a result of this I have to roll back to 4.15.0 on test runners. But that is no longer allowed by the installer, which says "Oh, it's up-to-date", meaning that I'll have to uninstall and reinstall to get to 4.15.0.

<img width="503" alt="tb-wsldd-05" src="https://user-images.githubusercontent.com/112444/213570710-26eb54c9-7098-44a9-9c07-2487ccaced82.png">

**Comment by rfay** [comment]
`docker context use default` works. 
`docker context inspect default` works.

**Comment by rfay** [comment]
And reverting to v4.15.0 I still have this problem, so I imagine it's some file left as-is during uninstall? Tried removing .docker/config.json but that didn't help.

**Comment by rfay** [comment]
I did find a 4.16.1 (that was updated from a previous version) that does not show this behavior. These two that are failing are fresh installs.

**Comment by rfay** [comment]
Now this is happening on a 3rd machine running 4.16.1, and I have no solutions.

**Comment by nicks** [comment]
Under the hood, I think 'docker context ls' reads json files on disk, decodes them, and pretty prints them it doesn't change much. I doubt it changed between 4.15 and 4.16. I think the files are under ~/.docker/contexts or something similar? Probably they're getting corrupted somehow by some command that runs before.

Maybe there's a docker/cli bug here to print the file where the error is coming from.

**Comment by rfay** [comment]
Thanks @nicks - 

Here's what I saw:
```
$ ls -l .docker/contexts/meta/*/*
-rwxrwxrwx 1 buildkite-agent buildkite-agent 0 Jan 19 16:31 .docker/contexts/meta/fe9c6bd7a66301f49ca9b6a70b217107cd1284598bfc254700c989b916da791e/meta.json
```

On each system there was a meta.json that was empty. Deleting them made `docker context ls` work.

**Comment by KarelVendla** [comment]
Any luck?

Also experiencing this problem :/

**Comment by nicks** [comment]
possibly related - https://github.com/docker/for-win/issues/12561 (though the exact symptoms are slightly different)

**Comment by rfay** [comment]
Thanks - these context files seem to be empty, so I think although the symptom is the same, the cause may be different.

**Comment by fari-99** [comment]
so i got this problem too, using Docker Desktop for Windows version 4.16.2 (95914).

the only fix i can do is using https://github.com/docker/desktop-linux/issues/57#issuecomment-1261852907 solution.

1. `vim ~/.docker/context/meta/{some hash}/meta.json`
2. put this in the meta.json
`{
  "Name": "default",
  "Metadata": {
    "StackOrchestrator": "swarm"
  },
  "Endpoints": {
    "docker": {
      "Host": "unix:///var/run/docker.sock",
      "SkipTLSVerify": false
    }
  },
  "TLSMaterial": {}
}`
3. do `docker context ls` again.

unfortunately, when i restart my docker desktop, the context got cleared again. and i need to do this all over again.
just a workaround, not a true solution.

**Comment by nicks** [comment]
Clarification question @rfay : after you delete the directory, how long does it take to re-occur? On every desktop restart? After a few hours? Intermittently?

**Comment by KarelVendla** [comment]
I am also facing this problem.

I've created a dummy context example:
`docker context create wsl-default`
And then deleted the broken one.

I get the issue less often, but it still occurs. Before I would just delete the hash folder with the empty meta.json, then the problem would come back a lot quicker (I recon docker doesn't like when your meta folder is empty).

It seems to occur when I crash my docker, it sometimes happens when I've restarted my PC. Haven't really noticed a distinct pattern yet.

The context with empty meta.json always seems to have the same hash, maybe that might give some clues.

**Comment by fari-99** [comment]
sorry to interrupt, i finally fixed it by deleting my `.docker` file on the Windows side (in `C:/Users/{Your Username}/.docker`), and restarting my docker.

NOTE: i'm not "deleting" it perse, just rename it to `.docker-old` just to be safe if error occurred, i can just redo the rename.

**Comment by KarelVendla** [comment]
> sorry to interrupt, i finally fixed it by deleting my `.docker` file on the Windows side (in `C:/Users/{Your Username}/.docker`), and restarting my docker.
> 
> NOTE: i'm not "deleting" it perse, just rename it to `.docker-old` just to be safe if error occurred, i can just redo the rename.

Do you mean .docker folder? I don't seem to have a .docker file

**Comment by fari-99** [comment]
> > sorry to interrupt, i finally fixed it by deleting my `.docker` file on the Windows side (in `C:/Users/{Your Username}/.docker`), and restarting my docker.
> > NOTE: i'm not "deleting" it perse, just rename it to `.docker-old` just to be safe if error occurred, i can just redo the rename.
> 
> Do you mean .docker folder? I don't seem to have a .docker file

yeah the .docker folder.

**Comment by rfay** [comment]
Deleting ~/.docker would fix it certainly until the next Docker Desktop restart. Same as deleting ~/.docker/contexts

**Comment by rfay** [comment]
@nicks if you delete ~/.docker/contexts it solves this problem until the next Docker Desktop restart, when the empty context json file immediately reappears and breaks `docker context list`.

**Comment by bplasmeijer** [comment]
Deleting ~/.docker/contexts works for me too.

How can this be fixed ASAP @thaJeztah ? Do we need to create a ticket at moby too?

**Comment by thaJeztah** [comment]
The `docker context` code is all client-side (https://github.com/docker/cli repository), but I don't think there's been changes in that area in 20.10.x patch releases of the cli (https://github.com/docker/cli/pulls?q=is%3Apr++is%3Aclosed+label%3Aarea%2Fcontext+)

On Docker Desktop, its _possible_ that other binaries interact with those files, e.g. the compose-cli wrapper (used for cloud integration) may also be interacting with them; https://github.com/docker/compose-cli/blob/7a57a330f6d94247e36dc868b3d9d7161fb4429f/api/context/store/store.go#L81-L97

Not sure when/in what cases it would though.

**Comment by rfay** [comment]
Thanks for looking at this @thaJeztah - Since restarting Docker Desktop (and not doing anything else) creates the empty context json files, it seems to implicate some kind of interaction.

**Comment by bplasmeijer** [comment]
Thanks for the feedback @thaJeztah 
Should I create a cross-link on the https://github.com/docker/cli repository?

**Comment by thaJeztah** [comment]
At this moment it's not 100% clear where the cause is (so it could be fully unrelated to the CLI code). Maybe @nicks has done more digging into the issue though (from the Docker Desktop side of things).

**Comment by crazy-max** [comment]
> The `docker context` code is all client-side (https://github.com/docker/cli repository), but I don't think there's been changes in that area in 20.10.x patch releases of the cli (https://github.com/docker/cli/pulls?q=is%3Apr++is%3Aclosed+label%3Aarea%2Fcontext+)
> 
> On Docker Desktop, its _possible_ that other binaries interact with those files, e.g. the compose-cli wrapper (used for cloud integration) may also be interacting with them; https://github.com/docker/compose-cli/blob/7a57a330f6d94247e36dc868b3d9d7161fb4429f/api/context/store/store.go#L81-L97
> 
> Not sure when/in what cases it would though.

That's odd, I can't find `failed to read metadata` on 20.10 branch but there is smth on master: https://github.com/docker/cli/blob/e92dd87c3209361f29b692ab4b8f0f9248779297/cli/context/store/metadatastore.go#L116

On buildx we are vendoring docker cli 23, that's might be why: https://github.com/docker/buildx/blob/cb94298a0238f67e008658e831af8b8dd313d444/go.mod#L10

I have no clue why compose users would have this issue though: https://github.com/docker/compose/blob/f24d3458c6d05e19519e660e4faa04d47f6db103/go.mod#L14

But best guess is cli vendored in buildx messed up the context store and therefore impacts compose after that.

**Comment by thaJeztah** [comment]
@crazy-max compose also has a replace rule https://github.com/docker/compose/blob/f24d3458c6d05e19519e660e4faa04d47f6db103/go.mod#L159

**Comment by crazy-max** [comment]
@thaJeztah Could be linked to https://github.com/docker/cli/pull/3790?

**Comment by crazy-max** [comment]
I think this could be the issue for Windows users: https://github.com/docker/cli/pull/3790/files#diff-60b97a3d3243f773d7c0d15a0d0d08d2019e75aabea298bf104094be66bf9047R150

**Comment by thaJeztah** [comment]
@crazy-max you mean https://github.com/docker/cli/commit/f843c42c05b7f50dc4204cf10379e45dca26c87a ? Trying to see the issue there; can you explain? 😅

**Comment by crazy-max** [comment]
> @crazy-max you mean [docker/cli@f843c42](https://github.com/docker/cli/commit/f843c42c05b7f50dc4204cf10379e45dca26c87a) ? Trying to see the issue there; can you explain? 😅

It seems only Windows users report this issue so far and with `filepath.Join` it would set `...\...` instead of `.../...`. Didn't go through the whole logic yet to understand where it's used but if it's an identifier in the store it might break?

**Comment by thaJeztah** [comment]
Ah, get your train of thought. Not sure that's it though (although that part of the code is somewhat debatable, as I think it was written to allow nested contexts ("who knows, maybe we need those at some point?"));

The results of that function is a list of paths;
https://github.com/docker/cli/blob/e92dd87c3209361f29b692ab4b8f0f9248779297/cli/context/store/metadatastore.go#L110-L111

Which get passed to `store.getByID()`, where "id" is effectively just the directory (don't ask), which gets `filepath.Join()`'d;

https://github.com/docker/cli/blob/e92dd87c3209361f29b692ab4b8f0f9248779297/cli/context/store/metadatastore.go#L67-L68

https://github.com/docker/cli/blob/e92dd87c3209361f29b692ab4b8f0f9248779297/cli/context/store/metadatastore.go#L25-L27

In either case, that would not explain how the empty JSON file would've been written. As far as I can tell (in the CLI implementation), only `createOrUpdate()` writes to those files;
https://github.com/docker/cli/blob/e92dd87c3209361f29b692ab4b8f0f9248779297/cli/context/store/metadatastore.go#L29-L39

And, at a glance, both "update" and "create" would never pass `nil` to that;
https://github.com/docker/cli/blob/e92dd87c3209361f29b692ab4b8f0f9248779297/cli/command/context/update.go#L97

https://github.com/docker/cli/blob/e92dd87c3209361f29b692ab4b8f0f9248779297/cli/command/context/create.go#L118

**Comment by thaJeztah** [comment]
So the only thing I'm wondering if it could be something odd with WSL2. I'm not too familiar with the internals/mechanisms WSL2 uses, but seem to recall the filesystem (or parts thereof) can be shared between the host (Windows) and WSL2? I did find some mention in the Docker Desktop source of an issue around some of that; @djs55 any ideas?

```go
func removeUnexpectedReparsePoint(path string) error {
	// POS-274: in 3.5.x there was a bug in the WSL 2 file sync logic triggered when the
	// Linux WSL homedir == the Windows homedir. The bug would leave various files inside
	// .docker as dangling WSL symlinks / Windows NTFS directory junctions.
	// We fix by moving these damaged files out of the way.
```

**Comment by zach-hill** [comment]
> So the only thing I'm wondering if it could be something odd with WSL2. I'm not too familiar with the internals/mechanisms WSL2 uses, but seem to recall the filesystem (or parts thereof) can be shared between the host (Windows) and WSL2? I did find some mention in the Docker Desktop source of an issue around some of that; @djs55 any ideas?
> 
> ```go
> func removeUnexpectedReparsePoint(path string) error {
> 	// POS-274: in 3.5.x there was a bug in the WSL 2 file sync logic triggered when the
> 	// Linux WSL homedir == the Windows homedir. The bug would leave various files inside
> 	// .docker as dangling WSL symlinks / Windows NTFS directory junctions.
> 	// We fix by moving these damaged files out of the way.
> ```

WSL2 mounts your windows filesystem using 9p at (by default) `/mnt/<drive_letter>` -- it seems in WSL2, my `~/.docker/contexts` folder is symlinked to my windows `~/.docker/contexts`., so if that's doing some kind of cleanup on symlinks, maybe that's what's happening? 

Something possibly to note about that code comment; both my windows user and WSL2 user share the same name(i.e. home path is `/home/zach` on WSL and `C:\users\zach` on windows), not sure if that's what it's referencing when it says `WSL homedir == Windows homedir`.

**Comment by bplasmeijer** [comment]
before docker restart
```
.
├── buildx
│   ├── current
│   ├── defaults
│   └── instances
├── config.json
├── contexts
│   └── meta
├── desktop
│   └── log
│       └── host
│           └── docker-desktop-user-distro.log
├── features.json -> /mnt/c/Users/bap/.docker/features.json
├── run
│   └── docker-cli-api.sock
└── scan
    └── config.json
```
after docker desktop restart
```
.
├── buildx
│   ├── current
│   ├── defaults
│   └── instances
├── config.json
├── contexts -> /mnt/c/Users/bap/.docker/contexts
├── desktop
│   └── log
│       └── host
│           └── docker-desktop-user-distro.log
├── features.json -> /mnt/c/Users/bap/.docker/features.json
├── run
│   └── docker-cli-api.sock
└── scan
    └── config.json
```
with symlink to local drive; both windows and wsl2
```
docker context ls
unexpected end of JSON input
```
remove contexts folder
both windows and wsl2 show
```
docker context ls
NAME                TYPE                DESCRIPTION                               DOCKER ENDPOINT               KUBERNETES ENDPOINT                                                                ORCHESTRATOR
default *           moby                Current DOCKER_HOST based configuration   unix:///var/run/docker.sock   https://bogus
```
meta file created by docker desktop restart
```
ls
~/.docker/contexts/meta/fe9c6bd7a66301f49ca9b6a70b217107cd1284598bfc254700c989b916da791e
cat meta.json
EMPTY
```

add PM WSL2 @craigloewen-msft and cross ticket to wsl https://github.com/microsoft/WSL/issues/9569

**Comment by thaJeztah** [comment]
@zach-hill @bplasmeijer thank you both for filling in some of the blanks for me! I'm checking up internally as well 👍

**Comment by bplasmeijer** [comment]
> @zach-hill @bplasmeijer thank you both for filling in some of the blanks for me! I'm checking up internally as well 👍

Also, thank you, @thaJeztah 
Expose the docker desktop codebase ;-) as requested many times ;-)

**Comment by thaJeztah** [comment]
> Expose the docker desktop codebase ;-) as requested many times ;-)

Ha! I have suggested that at times, but that's something _waaaay_ above my paygrade. I definitely continue trying to open-up code where possible, and for parts where that's not possible to be as _transparent_ as possible.

That said: in the unlikely event I become CEO: remind me to that

**Comment by rfay** [comment]
Just had this again on a fresh DD 4.16.3, and do in fact note that the in-wsl .docker/contexts is in fact a symlink to the Windows contexts directory, so Docker Desktop on the Windows side creating empty meta.json file will in fact bleed over into wsl2.

```
rfay@rfay-thinkbook:~/.docker$ ls -l contexts
lrwxrwxrwx 1 rfay rfay 35 Feb  1 10:31 contexts -> /mnt/c/Users/randy/.docker/contexts

rfay@rfay-thinkbook:~/.docker$ ls -l contexts
lrwxrwxrwx 1 rfay rfay 35 Feb  1 10:31 contexts -> /mnt/c/Users/randy/.docker/contexts
```

Also, deleting contexts directory on either side fixes things as you'd expect in this case.

**Comment by thaJeztah** [comment]
Thanks!

@nicks commented on our internal Slack thread that he managed to reproduce the issue (and suspects it may be something with WSL2). He's trying to get a consistent reproducer to trigger the issue (which would be the first step to (hopefully) fixing it, or at least narrowing down the cause).

**Comment by nicks** [comment]
ya, I saw this happen once this morning, but have been unable to reproduce it again. It happened when I was switching back and forth between hyper-v and wsl2.

I've been trying different combinations of reset / restart / 'docker context update' in parallel loops to probe race conditions, but without much success.

**Comment by bplasmeijer** [comment]
> ya, I saw this happen once this morning, but have been unable to reproduce it again. It happened when I was switching back and forth between hyper-v and wsl2.
> 
> I've been trying different combinations of reset / restart / 'docker context update' in parallel loops to probe race conditions, but without much success.

I can show you on my machine. My Story above valid all the time when I repeat, and repeat.

**Comment by nicks** [comment]
@bplasmeijer can you post a diagnostic id after it happens? https://docs.docker.com/desktop/troubleshoot/overview/#diagnose-from-the-terminal , run `com.docker.diagnose.exe gather -upload`

**Comment by rfay** [comment]
Just happened to me again, 300611F6-030B-45C1-8C80-A0A95C667D16/20230201182309

**Comment by nicks** [comment]
@rfay thanks! that one has a _particularly_ suspicious log of two commands in wsl interleaving, but i don't yet have a good mental model of why those two things interleaving might cause this problem, passing it along to some people who know more about the wsl interop if they have theories.

**Comment by nicks** [comment]
@rfay (or anyone else experiencing this issue frequently) - can you try this build and see if it helps?

https://desktop-stage.docker.com/win/main/amd64/97189/Docker%20Desktop%20Installer.exe

I made some changes to the concurrency of the context store that might make this much less likely to happen.
 
The platonic ideal fix would use something like https://github.com/golang/go/issues/33974 in docker/cli to put an advisory lock on the file, but that's a longer convo.

**Comment by bplasmeijer** [comment]
@nicks  https://github.com/docker/compose/issues/9956 ?
Do you want a log after a failed restart and docker context ls?

**Comment by nicks** [comment]
@bplasmeijer ya, it's helpful to have logs of the sequence:
1) `docker context ls` works
2) you restart DD
3) `docker context ls` fails with this error
because this sequence will tell us what's happening in DD when the file gets corrupted, and helps rule out possible codepaths where this could be happening.

or you could try the debug build in https://github.com/docker/for-win/issues/13180#issuecomment-1414184181 and see if it still re-corrupts the file.

**Comment by Yahav** [comment]
> (or anyone else experiencing this issue frequently) - can you try this build and see if it helps?

Is this safe to use though?

**Comment by schlich** [comment]
i've been running into this issue for weeks, also on WSL2, it's been a giant pain, especially when trying to boot up dev containers. I will give a go at trying that build! let me know if i can provide any helpful logs.

**Comment by schlich** [comment]
That debug build didn't work, i'm still getting blank generated `~/.docker/contexts/meta/<hash>/meta.json` files.

**Comment by schlich** [comment]
I was previously only attempting to clear context via clearing the .docker/contexts folder on the WSL side, but hadn't thought to check out the Windows side until i saw that comment upthread. On a hunch I erased the newline at the end of the meta.json on the Windows side, and I *think* that's fixed the issue?  🙏

**Comment by fari-99** [comment]
all delete `.docker/contexts` folder in WSL or in Windows is a temporary solution (at least for me). it will be back once in a while. don't know why tho. lmao

**Comment by nicks** [comment]
ok, i have a sandbox that reliably reproduces this, it's here:
https://github.com/nicks/contextstore-sandbox/blob/main/main.go

i think there are multiple levels of safety checks in docker/cli and docker/for-win that will help mitigate this, but the sandbox will help me test them.

**Comment by MarkWard0110** [comment]
I hit this problem today.  If it helps, here is a recap of my activity.
Working with VSCode dev containers.  I attempted to use the Dev Container command "Create Dev Container..." in a folder when it appeared the initialization stopped and would not continue.  Executing "docker" on the command line would execute but hang and not exit.  I opened ubuntu (WSL2), and it informed me that Microsoft Store now manages WSL, and I should execute `wsl.exe --upgrade` to link it.  I rebooted my computer and initiated the wsl.exe --upgrade.  After the upgrade I opened a folder in VSCode and executed the "Create Dev Container..." command, which opened up the selected dev container.  I repeated this activity for various different kinds of dev container templates.    
The area where I encountered a problem with dev containers was opening a folder having a .devcontainer directory.  I used the "Add Dev Container Configuration Files" command, and after selecting a template, I attempted to open the folder in the dev container.  This is when it would fail to open, and the dev container build would fail.  The "Create Dev Container..." command still worked.
I deleted the .docker/context directory, and I can open a folder having a .devcontainer directory in vscode dev container.

Maybe there is something different between the dev container "Create Dev Container" and the open using the local .devcontainer directory.

**Comment by DenyWatanabe** [comment]
Small contribution, in Windows if I edit ~/.docker/contexts/meta/<hash>/meta.json and change it to {}, I get:
![image](https://user-images.githubusercontent.com/25889678/218544035-f685a210-569d-4c08-926e-26f4928c9940.png)

Upon restarting docker, the contents of ~/.docker/contexts/meta/<hash>/meta.json are automatically changed to:
`{"Name":"desktop-linux","Metadata":{},"Endpoints":{"docker":{"Host":"npipe:////./pipe/dockerDesktopLinuxEngine","SkipTLSVerify":false}}}`

Also `docker context list` continues working exactly the same as above.

**Comment by alchatti** [comment]
@DenyWatanabe thanks your solution worked for me. I'm on W11 using WSL and the meta.json under my windows profile .docker folder `meta.json` was empty so I have added `{}`

`C:\Users\(USERNAME)\.docker\contexts\meta\(HASH)\meta.json`

(https://github.com/docker/for-win/issues/13180#issuecomment-1428456098)

**Comment by mdelgado** [comment]
hi! I hope my solution helps someone.
In my case I did as @alchatti suggested [here](https://github.com/docker/for-win/issues/13180#issuecomment-1429630445), but also a couple of symlinks were broken:
![image](https://user-images.githubusercontent.com/1399586/219455206-3249fe5a-b2e2-4cd3-9290-82bdb9d951f2.png)

Those two symlinks shown above were broken, so I releted them and recreated them, then applied the change that @alchatti suggested and restarted docker.

Hope it helps!

**Comment by radthenone** [comment]
> so i got this problem too, using Docker Desktop for Windows version 4.16.2 (95914).
> 
> the only fix i can do is using [docker/desktop-linux#57 (comment)](https://github.com/docker/desktop-linux/issues/57#issuecomment-1261852907) solution.
> 
> 1. `vim ~/.docker/context/meta/{some hash}/meta.json`
> 2. put this in the meta.json
>    `{   "Name": "default",   "Metadata": { "StackOrchestrator": "swarm"   },   "Endpoints": { "docker": { "Host": "unix:///var/run/docker.sock", "SkipTLSVerify": false }   },   "TLSMaterial": {} }`
> 3. do `docker context ls` again.
> 
> unfortunately, when i restart my docker desktop, the context got cleared again. and i need to do this all over again. just a workaround, not a true solution.

thanks man

**Comment by lainosantos** [comment]
Hi there.

Only deleting the  folder %USERPROFILE%\.docker\contexts\meta on Windows side solves the problem even after restart the docker desktop.

**Comment by radthenone** [comment]
I delete .docker, close IDE, off docker-desktop, restart PC, and now all works and now i dont need reopen meta file.

**Comment by chaizeg** [comment]
Closing this issue because a fix has been released in Docker Desktop `4.17.0` . See the [release notes](https://docs.docker.com/desktop/release-notes/) for more details.
